src/minmax.rs

```diff
@@ -0,0 +1,281 @@
+        val.time = Some(time);
+        val.value = meas;
+
+        let temp = val
+            .time
+            .unwrap()
+            .duration_since(self.estimate[2].time.unwrap());
+
+        // if nothing in window or found new max value, reset!
+        if val.value.ge(&self.estimate[0].value) || temp.gt(&win) {
+            return self.reset(time, meas);
+        }
+
+        if val.value.ge(&self.estimate[1].value) {
+            self.estimate[2] = val;
+            self.estimate[2] = val;
+        } else if val.value.ge(&self.estimate[2].value) {
+            self.estimate[2] = val
+        }
+
+        self.subwin_update(win, time, meas)
+    }
+
+    fn subwin_update(
+        &mut self, win: Duration, time: Instant, meas: Duration,
+    ) -> Duration {
+        let mut val = MinmaxSample::default();
+        val.time = Some(time);
+        val.value = meas;
+
+        let dt = time.duration_since(self.estimate[0].time.unwrap());
+
+        if dt.gt(&win) {
+            self.estimate[0] = self.estimate[1];
+            self.estimate[1] = self.estimate[2];
+            self.estimate[2] = val;
+
+            let temp = time.duration_since(self.estimate[0].time.unwrap());
+            if temp.gt(&win) {
+                self.estimate[0] = self.estimate[1];
+                self.estimate[1] = self.estimate[2];
+                self.estimate[2] = val;
+            }
+        } else if self.estimate[1]
+            .time
+            .unwrap()
+            .eq(&self.estimate[0].time.unwrap()) &&
+            dt.gt(&win.div_f32(4.0))
+        {
+            self.estimate[2] = val;
+            self.estimate[1] = val;
+        } else if self.estimate[2]
+            .time
+            .unwrap()
+            .eq(&self.estimate[1].time.unwrap()) &&
+            dt.gt(&win.div_f32(2.0))
+        {
+            self.estimate[2] = val;
+        }
+        self.estimate[0].value
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn reset_filter() {
+        let mut f = Minmax::default();
+        let now = Instant::now();
+        let rtt = Duration::new(0, 50000000);
+
+        let rtt_min = f.reset(now, rtt);
+        assert_eq!(rtt_min, rtt);
+
+        assert_eq!(f.estimate[0].time, Some(now));
+        assert_eq!(f.estimate[0].value, rtt);
+
+        assert_eq!(f.estimate[1].time, Some(now));
+        assert_eq!(f.estimate[1].value, rtt);
+
+        assert_eq!(f.estimate[2].time, Some(now));
+        assert_eq!(f.estimate[2].value, rtt);
+    }
+
+    #[test]
+    fn get_windowed_min() {
+        let mut f = Minmax::default();
+        let rtt_25 = Duration::new(0, 25000000);
+        let rtt_24 = Duration::new(0, 24000000);
+        let win = Duration::new(1, 0);
+
+        let mut rtt_min = f.reset(Instant::now(), rtt_25);
+        assert_eq!(rtt_min, rtt_25);
+
+        std::thread::sleep(Duration::new(0, 500000000));
+        rtt_min = f.running_min(win, Instant::now(), rtt_24);
```

Comment:
Can you update not to use `sleep()`? We can use now + Duration(x) to simulate some time in the future.